text/0033-crash-recovery.md

```diff
@@ -0,0 +1,108 @@
+
+For the units, it is enough to save the hash of the latest created unit. This would make it possible to download the unit from other validators along with its whole history and be able to verify the correctness of downloaded data.
+
+Units aren't the only thing that might be used to create an equivocation in Highway, though. Endorsements are such a thing, too. It would therefore be necessary to save the hashes of all endorsed units, too. When rejoining the era, the validator would read the list of such hashes, download the relevant units and make sure that it never endorses conflicting units in the future.
+
+The only remaining detail is how to implement saving the relevant hashes in practice. The simplest solution of just saving the hash whenever we create a unit or endorsement has a significant drawback. Imagine we create a unit, save its hash as the hash of the last unit created by us, and then crash before sending the unit to any validator (for any reason, eg. a power outage). In such a case, after restarting, we wouldn't be able to download it again. We wouldn't be able to make sure that we don't equivocate and so we wouldn't start voting again. It is probably the most secure solution, though. An alternative will be discussed in [Rationale and alternatives](#rationale-and-alternatives).
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The consensus component should:
+
+- Whenever a unit or an endorsement is created, emit a request to storage to store the latest data for the era. The data stored would contain:
+    - The hash of the latest unit
+    - The list of hashes of latest units by each validator endorsed by us
+
+This data should be keyed by the consensus instance ID.
+
+In order to facilitate removing data for old eras (ones that are older than `UNBONDING_PERIOD`), we should also store a map of era ID → instance ID - as instance IDs themselves carry no information about how old the era is. We can, however, identify the era IDs of old eras, and using this map, we could also find the entries to be removed.
+
+Any sending of a unit / an endorsement should only happen after the storage request returns a success response.
+
+- Whenever a new era is started, it should first try to read the entry with latest hashes from storage by submitting a read request.
+- If the entry doesn't exist, proceed the same way it is done currently.
+- If the entry exists:
+    - Pass all the data to the `ActiveValidator` instance.
+    - Only create new units if the unit with the latest hash is known to the consensus state.
+    - Only create new endorsements if all the previous endorsements are known to the consensus state.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+Every creation of a unit or an endorsement will now involve storage - however, this seems unavoidable in that we need to store the knowledge of our activity in consensus in some permanent way.
```

Comment:
I think we need to mention that if a node crashes between persisting the new unit and gossiping it, then upon restart it won't be able to join that era.